roachprod: auto restart option for cockroach processes

herkolategan opened the pull request:
Previously, if a cockroach process died, it needed to be manually restarted.
This is expected in most circumstances. But this behavior might not always be
practical, for instance, when running a long unattended benchmark.

To support this, a new start option has been added that automatically restarts
the service if a process fails (e.g., due to a disk stall). This ensures the
process comes back online quickly without manual intervention.

Epic: None
Release note: None

herkolategan committed:
roachprod: auto restart option for cockroach processes

Previously, if a cockroach process died, it needed to be manually restarted.
This is expected in most circumstances. But this behavior might not always be
practical, for instance, when running a long unattended benchmark.

To support this, a new start option has been added that automatically restarts
the service if a process fails (e.g., due to a disk stall). This ensures the
process comes back online quickly without manual intervention.

Epic: None
Release note: None

herkolategan committed:
roachprod: add auto restart option flag

Add a flag to the roachprod CLI to enable auto restarting of cockroach
processes.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154383)
<!-- Reviewable:end -->

srosenberg commented on pkg/cmd/roachprod/cli/flags.go:
```diff
@@ -453,6 +453,8 @@ func initFlagsStartOpsForCmd(cmd *cobra.Command) {
 		"enable-fluent-sink", startOpts.EnableFluentSink,
 		"whether to enable the fluent-servers attribute in the CockroachDB logging configuration")
+	cmd.Flags().BoolVar(&startOpts.AutoRestart,
+		"auto-restart", startOpts.AutoRestart, "automatically restart any process that dies")
```
Nit: `any` -> `cockroach`

herkolategan committed:
roachprod: add burst and interval limit to auto restart

Adding burst and interval limit to auto restart in order to prevent an infinite
restart (crash) loop,

Epic: None
Release note: None

herkolategan commented:
TFTR!

bors r=srosenberg
